src/object/srv_ec_aggregate.c

```diff
@@ -0,0 +1,2220 @@
+			 agg_recx_iter_pre_cb, NULL, entry, NULL);
+	/* entry->ae_par_extent.ape_epoch has been set to the parity extent's
+	 * epoch
+	 */
+	if (rc != 0)
+		goto out;
+
+	D_DEBUG(DB_TRACE, "Par query: epoch: %lu, offset: %lu, length: %lu\n",
+		entry->ae_par_extent.ape_epoch,
+		entry->ae_par_extent.ape_recx.rx_idx,
+		entry->ae_par_extent.ape_recx.rx_nr);
+
+	/* change to >= to deal with retained extent */
+	if (entry->ae_par_extent.ape_epoch >= entry->ae_cur_stripe.as_hi_epoch
+	    && entry->ae_par_extent.ape_epoch != ~(0ULL)) {
+		/* Parity newer than data and holes; so we need to delete
+		 * the replicas!
+		 */
+		update_vos = false;
+		goto out;
+	}
+
+	if ((entry->ae_par_extent.ape_epoch == ~(0ULL)
+				 && agg_stripe_is_filled(entry, false)) ||
+					agg_stripe_is_filled(entry, true)) {
+		/* Replicas constitute a full stripe. */
+		rc = agg_encode_local_parity(entry);
+		goto out;
+	}
+	if (entry->ae_par_extent.ape_epoch == ~(0ULL)) {
+		/* No parity, partial-stripe worth of replicaa, nothing to do */
+		update_vos = false;
+		goto out;
+	}
+	/* Parity, some later replicas, possibly holes, not full stripe. */
+	if (entry->ae_cur_stripe. as_has_holes)
+		process_holes = true;
+	else
+		rc = agg_process_partial_stripe(entry);
+
+out:
+	if (process_holes && rc == 0)
+		rc = agg_process_holes(entry);
+	else if (update_vos && rc == 0) {
+		entry->ae_cur_stripe.as_suffix_ext = agg_get_carry_under(entry);
+		if (!rc && entry->ae_oca->u.ec.e_p > 1) {
+			/* offload of ds_obj_update to push remote parity */
+			rc = agg_peer_update(entry);
+			if (rc)
+				D_ERROR("agg_peer_update fail: "DF_RC"\n",
+					DP_RC(rc));
+		}
+		if (rc == 0) {
+			rc = agg_update_vos(entry);
+			if (rc)
+				D_ERROR("agg_update_vos fail: "DF_RC"\n",
+					DP_RC(rc));
+		}
+	}
+
+	agg_clear_extents(entry);
+	return rc;
+}
+
+/* Returns the subrange of the RECX iterator's returned recx that lies within
+ * the current stripe.
+ */
+static daos_off_t
+agg_in_stripe(struct ec_agg_entry *entry, daos_recx_t *recx)
+{
+	unsigned int		len = entry->ae_oca->u.ec.e_len;
+	unsigned int		k = entry->ae_oca->u.ec.e_k;
+	daos_off_t		stripe = recx->rx_idx / (len * k);
+	daos_off_t		stripe_end = (stripe + 1) * len * k;
+
+	if (recx->rx_idx + recx->rx_nr > stripe_end)
+		return stripe_end - recx->rx_idx;
+	else
+		return recx->rx_nr;
+}
+
+/* Iterator call back sub-function for handling data extents.
+ */
+static int
+agg_data_extent(vos_iter_entry_t *entry, struct ec_agg_entry *agg_entry,
+		daos_handle_t ih, unsigned int *acts)
+{
+	struct ec_agg_extent	*extent = NULL;
+	int			 rc = 0;
+
+	if (agg_stripenum(agg_entry, entry->ie_recx.rx_idx) !=
+			agg_entry->ae_cur_stripe.as_stripenum) {
+		/* Iterator has reached next stripe */
+		if (agg_entry->ae_cur_stripe.as_stripenum != ~0UL) {
+			rc = agg_process_stripe(agg_entry);
+				D_ERROR("Process stripe returned "DF_RC"\n",
+					DP_RC(rc));
+		/* Error leaves data covered by replicas vunerable to vos delete
+		 */
+		 	rc = 0;
```

Comment:
(style)  code indent should use tabs where possible
(style)  please, no space before tabs